sycl/include/sycl/reduction.hpp

```diff
@@ -1124,11 +1124,58 @@ void doTreeReductionHelper(size_t WorkSize, size_t LID, BarrierTy Barrier,
   }
 }
 
-template <typename LocalRedsTy, typename BinOpTy, typename BarrierTy>
-void doTreeReduction(size_t WorkSize, size_t LID, LocalRedsTy &LocalReds,
-                     BinOpTy &BOp, BarrierTy Barrier) {
+// Enum for specifying work size guarantees in tree-reduction.
+enum WorkSizeGuarantees { None, Equal, LessOrEqual };
+
+template <WorkSizeGuarantees WSGuarantee, int Dim, typename LocalRedsTy,
+          typename BinOpTy, typename BarrierTy, typename AccessFuncTy>
+void doTreeReduction(size_t WorkSize, nd_item<Dim> NDIt, LocalRedsTy &LocalReds,
+                     BinOpTy &BOp, BarrierTy Barrier, AccessFuncTy AccessFunc) {
+  size_t LID = NDIt.get_local_linear_id();
+  size_t AdjustedWorkSize;
```

Comment:
We can also do something like this:

```
  auto AccessFuncCaller = [&](size_t I) {
    if constexpr (std::is_invocable_v<AccessFuncTy>)
      return AccessFunc();
    else
      return AccessFunc(I);   
  };
```

and cleanup `[&](size_t) -> [&]()` in callers.

Reply:
I am not convinced the additional complexity in `doTreeReduction` is worth it just to remove that ignored argument.